fastwarc/fastwarc/cli.py

```diff
@@ -167,13 +168,98 @@ def check(infile, decompress_alg, verify_payloads, quiet, output):
                     click.echo(rec)
                 sys.exit(1)
 
+@main.command()
+@click.argument('infile', type=click.Path(dir_okay=False, exists=True))
+@click.argument('offset', type=int)
+@click.option('-o', '--output', type=click.File('wb'), default=sys.stdout.buffer,
+              help='Output file, default is stdout')
+@click.option('--payload', is_flag=True,
+              help='Output only record payload (transfer and/or content encoding are preserved')
+@click.option('--headers', is_flag=True, help='Output only record (and HTTP) headers')
+def extract(infile, offset, output, payload, headers):
+    """Extract WARC record by offset."""
+    compl_record = not (payload or headers)
+    with open(infile, 'rb') as stream:
+        stream.seek(offset)
+        try:
+            record = next(ArchiveIterator(stream))
+            if compl_record:
+                record.write(output)
+            elif headers:
+                record.headers.write(PythonIOStreamAdapter(output))
+                output.write(b'\r\n')
+                if record.is_http:
+                    record.http_headers.write(PythonIOStreamAdapter(output))
+                    output.write(b'\r\n')
+            elif payload:
+                buf = record.reader.read(4096)
+                while buf:
+                    output.write(buf)
+                    buf = record.reader.read(4096)
+        except Exception as e:
+            sys.stderr.write('Failed to extract WARC record at offset %d: %s\n' % (offset, e))
```

Comment:
Thanks for adding this, though you should be more explicit. StopIteration is not something you would want to print (and it doesn't have an error message anyway). Instead, you would simply return an empty output.

Other exceptions you actually might want to print would be stream_io.StreamError and (theoretically) OSError and ValueError, which may be raised by Python stream objects.

Oh, and no legacy format strings, please. Use f-strings instead. ;-)